src/coreComponents/constitutive/fluid/PVTFunctions/CO2SolubilityFunction.cpp

```diff
@@ -274,121 +199,128 @@ void CO2SolubilityFunction::makeTable( string_array const & inputPara )
+void CO2Solubility::calculateCO2Solubility( array1d< array1d< real64 > > const & coordinates,
+                                            real64 const & salinity,
+                                            array1d< real64 > const & values )
 {
+  // Interaction parameters, see Table 2 of Duan and Sun (2003)
+  constexpr real64 mu[] =
+  { 28.9447706, -0.0354581768, -4770.67077, 1.02782768e-5, 33.8126098, 9.04037140e-3,
+    -1.14934031e-3, -0.307405726, -0.0907301486, 9.32713393e-4, 0 };
+  constexpr real64 lambda[] = { -0.411370585, 6.07632013e-4, 97.5347708, 0, 0, 0, 0, -0.0237622469, 0.0170656236, 0, 1.41335834e-5 };
+  constexpr real64 zeta[] = { 3.36389723e-4, -1.98298980e-5, 0, 0, 0, 0, 0, 2.12220830e-3, -5.24873303e-3, 0, 0 };
 
-  EvalArgs2D P, T, solubility;
-  P.m_var = pressure.m_var;
-  P.m_der[0] = 1.0;
+  localIndex const numPressures = coordinates[0].size();
+  localIndex const numTemperatures = coordinates[1].size();
 
-  T.m_var = temperature.m_var;
-  T.m_der[1] = 1.0;
+  for( localIndex i = 0; i < numPressures; ++i )
+  {
+    real64 const P = coordinates[0][i] / P_Pa_f;
 
-  //solubiltiy mol/kg(water)  X = Csat/W
-  solubility = m_CO2SolubilityTable->value( P, T );
+    for( localIndex j = 0; j < numTemperatures; ++j )
+    {
+      real64 const T = coordinates[1][j];
 
-  real64 const waterMW = m_componentMolarWeight[m_waterIndex];
+      // compute reduced volume by solving the CO2 equation of state
+      real64 V_r = 0.0;
+      CO2SolubilityFunction( T, P, V_r, &detail::ff );
 
-  solubility *= waterMW;
+      // compute equation (6) of Duan and Sun (2003)
+      real64 const logK = detail::Par( T+T_K_f, P, mu )
+                          - detail::logF( T, P, V_r )
+                          + 2*detail::Par( T+T_K_f, P, lambda ) * salinity
+                          + detail::Par( T+T_K_f, P, zeta ) * salinity * salinity;
 
-  EvalVarArgs X, Y;
+      // mole fraction of CO2 in vapor phase, equation (4) of Duan and Sun (2003)
+      real64 const y_CO2 = (P - detail::PWater( T ))/P;
+      values[j*numPressures+i] = y_CO2 * P / exp( logK );
+    }
+  }
+}
 
-  X.m_var = solubility.m_var;
-  X.m_der[0] = solubility.m_der[0];
+void CO2Solubility::CO2SolubilityFunction( real64 const & T,
+                                           real64 const & P,
+                                           real64 & V_r,
+                                           real64 (*f)( real64 const & x1, real64 const & x2, real64 const & x3 ) )
+{
 
-  //Y = C/W = z/(1-z)
+  constexpr real64 eps = 1e-9;
+  constexpr real64 dx = 1e-10;
+  int count = 0;
+  real64 Vr_int = 0.05;
 
-  if( compFraction[m_CO2Index].m_var > 1.0 - minForDivision )
-  {
-    Y = compFraction[m_CO2Index] / minForDivision;
-  }
-  else
-  {
-    Y = compFraction[m_CO2Index] / (1.0 - compFraction[m_CO2Index]);
-  }
+  V_r = 0.75*Rgas*(T_K_f+T)/(P*P_Pa_f)*(1/V_c);
 
-  if( Y < X )
+  // iterate until the solution of the CO2 equation of state is found
+  for(;; )
```

Comment:
Yes I was reluctant to changing that because I was afraid of breaking things (the equation of state solved in this loop seems quite nasty). I may try to remove it at the very end